### ISSUE TITLE
Support Composite SLO in OpenSLO.

### ISSUE THREAD

**Issue by davex98** [OpenSLO/OpenSLO#142]
### Problem to solve

Support Composite SLO in OpenSLO.
The goal of Composite SLO is to enable the user to capture an end-to-end journey.

### Proposal

The Composite SLO will provide users the ability to define an SLO that is based on the health of many independent objectives that are defined with different queries, targets, etc. (even data sources).
 
As an example, the banking industry normally has 10-20 services for a user scenario. Withdrawing money from an ATM requires multiple services and transactions to happen in low latency, high availability environment. The SLO in this case is the success of a customer withdrawing money from an ATM. One might imagine we would aim for 99.99% reliability and with all the underlying services and their SLOs making up the Composite SLO.
 
To enable the creation of a Composite SLO we will have to modify the current structure of our YAML files to enable specifying different SLI for each objective instead of having one SLI for one SLO object. We could achieve that by adding **indicatorRef:**
```yaml
objectives:
  - displayName: string # optional
    op: lte | gte | lt | gt # conditional operator used to compare the SLI against the value. Only needed when using a thresholdMetric
    value: numeric # optional, value used to compare threshold metrics. Only needed when using a thresholdMetric
    target: numeric [0.0, 1.0) # budget target for given objective of the SLO
    timeSliceTarget: numeric (0.0, 1.0] # required only when budgetingMethod is set to TimeSlices
    timeSliceWindow: number | duration-shorthand # required only when budgetingMethod is set to TimeSlices
    indicatorRef: name of the SLI
```

or we could simple inline our SLI in the objective:
```yaml
objectives:
  - displayName: string # optional
    op: lte | gte | lt | gt # conditional operator used to compare the SLI against the value. Only needed when using a thresholdMetric
    value: numeric # optional, value used to compare threshold metrics. Only needed when using a thresholdMetric
    target: numeric [0.0, 1.0) # budget target for given objective of the SLO
    timeSliceTarget: numeric (0.0, 1.0] # required only when budgetingMethod is set to TimeSlices
    timeSliceWindow: number | duration-shorthand # required only when budgetingMethod is set to TimeSlices
    thresholdMetric:
      metricSource: string
        metricSourceRef: string
        spec:
         # arbitrary chosen fields for every data source type to make it comfortable to use

```
### Further details

We could decide on our upcoming community meeting and discuss this way of handling Composite SLO in our standard.

**Comment by davex98** [comment]
> I'm definitely on board with the need for a way to create composite/aggregate SLOs for tracking user journey across services.
> 
> A related issue is #41
> 
> I can see a need for varied approaches on how the SLOs are aggregated into a single SLO, such as average of SLOs or taking a "one failed they all failed" type approach (an aggregate SLO with 3 sub SLOs is considered failing if just one of the sub SLOs is failing).
> 
> From a user experience perspective, I wonder if we need a new `SLOComposite` type that can use references to refer to SLOs and all the pieces, and the `SLOComposite` only defines the computation of how they're combined. While the above proposal keeps it simple by utilizing existing objects, I would have concerns related to the usability and possible confusion of whether an SLI should be off the root SLO object, or within the objectives, with this approach.
> 
> I couldn't think of it during today's meeting, but it could be interesting to explore the idea of a Synthetic SLI consisting of SLOs mentioned in #41

The approach of "one failed they all failed" seems reasonable, I think there might be various ways of calculating the aggregated SLO - how the error budget will be burned depending on the input. 

I have also a feeling that adding a new kind of `SLOComposite` could add extra complexity at this point in our specification. 
When thinking about moving `SLI` from the root of the `SLO` to the objectives the strong advantage was letting the user specify different queries/data sources within a single `SLO` and then we could add a new field in the spec that would indicate that a particular SLO is a Composite.

**Comment by kenfinnigan** [comment]
It may not change how we define it, but I would like to get clarification on what exactly we're doing a composite of.

Is the composite SLO proposal wanting the SLI to contain metrics, SLOs (where the SLI is utilizing a separate SLO), or some combination of both?

**Comment by programmer04** [comment]
Let's make it as flexible as needed. But referencing the whole SLO may be tricky because SLO can have multiple basically independent objectives thus we don't have one output of the SLO. Defining composite makes this type of aggregation

**Comment by kenfinnigan** [comment]
I was thinking on this problem earlier this week and we discussed it some more in yesterday's community meeting.

My current thinking is there are, at least, two distinct use cases we're trying to solve:

1. Multiple SLIs on a single SLO. This supports customer journey definitions where an SLI can represent a step in that journey, and also enabling each SLI to possibly be retrieved from different data sources.
2. Combining multiple SLOs to provide an overall perspective of a service across all its SLOs to indicate is it healthy or not

To support 1), we would also need the ability for an objective target to include a calculation on how the different SLIs should be combined to create a single measure the target must be compared against. This could include simple averages or weighted importance of each SLI being combined.

To support 2), as we're dealing with SLOs, which have targets, this combination can likely only have a simple calculation to determine healthy based on how many SLOs within the combination are breaching at a point in time. For example, if one is failing then fail, or > 50% of the SLOs are failing then fail, etc.

For 2), we likely need a different name as it's not really an SLO in the true sense as it does not have its own target or time window. Maybe `RollUpSLO`, or something similar, as a name?

Happy to open separate issues for any piece of this if needed.

Be interested to hear what others think

**Comment by fpiwowarczyk** [comment]
Resolved by #182
